Fix QUIC unit test failures

maskit opened the pull request:
Recent BoringSSL requires ALPN extension if the SSL context is used for QUIC.
https://boringssl.googlesource.com/boringssl/+/74161f485b5d54fe963cbd3d081b718ec84d2e00%5E%21/

bneradt commented on iocore/net/quic/test/test_QUICHandshakeProtocol.cc:
```diff
@@ -102,16 +102,32 @@ TEST_CASE("QUICHandshakeProtocol")
+  SSL_CTX_set_alpn_select_cb(
+    server_ssl_ctx,
+    [](SSL *ssl, const unsigned char **out, unsigned char *outlen, const unsigned char *in, unsigned inlen, void *) {
+      auto ret = SSL_select_next_proto(const_cast<unsigned char **>(out), outlen,
+                                       reinterpret_cast<const unsigned char *>("\6h3-foo"), 6, in, inlen);
```
The openssl docs aren't clear on this, but I believe the server_len parameter here should be 7 instead of 6. The prefix of '\6' in the server string is correct, but I think that the server_len has to count all characters in the server string, inclusive of any prefixes.

I think I'm reading the code right:
https://github.com/openssl/openssl/blob/3f71add9e57fb48cb5efdc765860daf754db40e9/ssl/ssl_lib.c#L2896

It will need to loop over all characters in server and client.

bneradt commented on iocore/net/quic/test/test_QUICPacketHeaderProtector.cc:
```diff
@@ -65,6 +65,20 @@ TEST_CASE("QUICPacketHeaderProtector")
+  SSL_CTX_set_alpn_select_cb(
+    server_ssl_ctx,
+    [](SSL *ssl, const unsigned char **out, unsigned char *outlen, const unsigned char *in, unsigned inlen, void *) {
+      auto ret = SSL_select_next_proto(const_cast<unsigned char **>(out), outlen,
+                                       reinterpret_cast<const unsigned char *>("\6h3-foo"), 6, in, inlen);
```
Same comment as above.

bneradt reviewed (changes requested):
Looks good. I just have a question about the server_len value.

maskit commented:
@bneradt Good catch. You're right. I've updated the lengths.
The protocol name was originally "h3-29" and I changed it to "h3-foo", and forgot to update the length.

bneradt reviewed (approved):
Looks good.

bneradt commented:
[approve ci autest]

zwoop commented:
Cherry-picked to v9.1.x branch.
